### PR TITLE
[SPARK-24468][SQL] Handle negative scale when adjusting precision for decimal operations

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

In SPARK-22036 we introduced the possibility to allow precision loss in arithmetic operations (according to the SQL standard). The implementation was drawn from Hive's one, where Decimals with a negative scale are not allowed in the operations.

The PR handles the case when the scale is negative, removing the assertion that it is not.

## How was this patch tested?

added UTs

### CODE REVIEW

**Comment by mgaido91** [comment]
cc @cloud-fan @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala [hunk 1]
```diff
@@ -161,13 +161,17 @@ object DecimalType extends AbstractDataType {
...
+    // Assumption:
     assert(precision >= scale)
-    assert(scale >= 0)
 
     if (precision <= MAX_PRECISION) {
       // Adjustment only needed when we exceed max precision
       DecimalType(precision, scale)
+    } else if (scale < 0) {
+      // Decimal can have negative scale (SPARK-24468). In this case, we cannot allow a precision
+      // loss since we would cause a loss of digits in the integer part.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala]: > cannot allow ...

do you mean we should return null for this case?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala]: yes, in this case we return `null` at the moment. I have a long standing PR to make this behavior configurable in order to be compliant to SQL standard (https://github.com/apache/spark/pull/20350), but it has not been very active lately....

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala]: is there a SQL standard for it? I feel it seems reasonable to truncate the integral part, like `123456` -> `123000`

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala]: Unfortunately, the SQL standard says nothing about this. It only states:

> the result is of a numeric type that depends only on the numeric type of the operands. If the result of an arithmetic operation cannot be represented exactly in its result type, then whether the result is rounded or truncated is implementation-defined. An
exception condition is raised if the result is outside the range of numeric values of the result type,

It says nothing about which should be the result scale/precision (it just says that it has to be based on the input types). So it is up to us to decide which approach to follow (ie. whether to be more aggressive in allowing precision loss or not).

The reasons of my choice are:
 - In the current implementation, when we allow precision loss we anyway enforce at least 6 decimal digits precision. So we are guaranteeing to our users that any (successful) arithmetical operation will have at least 6 decimal digits of precision anyway.  It'd be very weird to me that we do not allow to loose precision for operations which would be representable with 4 decimal digits, but we allow data loss on the integer part....
 - Hive and SQLServer on which we base our implementation do not allow negative scale, so we are already allowing to represent more operations than them. And they do not therefore allow any precision loss on the integer part;
 - I found no other RDBMS allowing a precision loss on the integer part.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala]: ok makes sense, do we have an end-to-end test case for returning null?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala]: I am adding it, thanks.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.3!

**Comment by gatorsmile** [comment]
Thank you for fixing this! In Spark 3.0, we might introduce a mode for strict ANSI compliance. We can handle the cases in https://github.com/apache/spark/pull/20350

**Comment by mgaido91** [comment]
sure, thank you @gatorsmile

**Pull request closed without merging**